sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -3894,16 +3899,99 @@ pi_result piextUSMSharedAlloc(void **ResultPtr, pi_context Context,
   ZeDevDesc.ordinal = 0;
   ZE_CALL(zeDriverAllocSharedMem(Context->Device->Platform->ZeDriver,
                                  &ZeDevDesc, &ZeHostDesc, Size, Alignment,
-                                 Device->ZeDevice, ResultPtr));
+                                 Context->Device->ZeDevice, ResultPtr));
 
   return PI_SUCCESS;
 }
 
-pi_result piextUSMFree(pi_context Context, void *Ptr) {
+pi_result USMFreeImpl(pi_context Context, void *Ptr) {
   ZE_CALL(zeDriverFreeMem(Context->Device->Platform->ZeDriver, Ptr));
   return PI_SUCCESS;
 }
 
+// Exception type to pass allocation errors
+class UsmAllocationException {
+  const pi_result Error;
+public:
+  UsmAllocationException(pi_result Err): Error{Err} {}
+  pi_result getError() const { return Error; }
+};
+
+void* USMSharedMemoryAlloc::allocate(size_t Size) {
+  void* Ptr = nullptr;
+
+  // TODO: how to deal with properties? Allocate a specific slab or just bypass the allocator?
+  // They seems to be not used by DPC++ RT right now.
+  auto Res = USMSharedAllocImpl(&Ptr, Context, nullptr, Size, sizeof(void*));
+  if (Res != PI_SUCCESS) {
```

Comment:
Could we stick to the way the rest of PI API interfaces are done, i.e. they always return pi_result (no C++ exceptions), and any out arguments are through a pointer. You will also need less code this way as a bonus.

Reply:
USMSharedMemoryAlloc and its methods are only called by the allocator itself, so it's not a part of PI API.
The problem here is that we can detect allocation failure, i.e. non success result, only on this lower layer. Using exception here make things easier because it eliminates the need to handle and propagate allocation error code through the allocator to the upper layer.

Reply:
I understand they are not a part of PI API. What I suggest is to follow PI API style of handling errors, i.e. having it as a return value for shortness and for being consistent with the rest of the code in L0 plugin.

Reply:
Is it just a consistency issue or is there is some technical limitation to use exceptions? If it's just due to consistency, can we make an exception here?
Using error code propagation instead of exception has the following issues:
- it requires additional checks put into allocator core logic(usm_allocator.cpp file) which makes it more complex.
- it leaks some memory allocation details into the allocator core 
I think it's worth to add some inconsistency here just to make the implementation less complex.

Reply:
Yes, we can leave at as is for now. There is no functional issue that I can see.